pkg/cmd/cli/cmd/volume.go

```diff
@@ -0,0 +1,588 @@
+	errList := []error{}
+	if isAddOp {
+		if len(a.Type) == 0 && len(a.Source) == 0 {
+			errList = append(errList, errors.New("must provide --type or --source for --add operation"))
+		} else if len(a.Type) > 0 && len(a.Source) > 0 {
+			errList = append(errList, errors.New("either specify --type or --source but not both for --add operation"))
+		}
+		if len(a.Type) > 0 {
+			switch strings.ToLower(a.Type) {
+			case "emptydir":
+			case "hostpath":
+				if len(a.Path) == 0 {
+					errList = append(errList, errors.New("must provide --path for --type=hostPath"))
+				}
+			case "secret":
+				if len(a.SecretName) == 0 {
+					errList = append(errList, errors.New("must provide --secret-name for --type=secret"))
+				}
+			case "persistentvolumeclaim":
+				if len(a.ClaimName) == 0 {
+					errList = append(errList, errors.New("must provide --claim-name for --type=persistentVolumeClaim"))
+				}
+			default:
+				errList = append(errList, errors.New("invalid volume type. Supported types: emptyDir, hostPath, secret, persistentVolumeClaim"))
+			}
+		} else if len(a.Path) > 0 || len(a.SecretName) > 0 || len(a.ClaimName) > 0 {
+			errList = append(errList, errors.New("--path|--secret-name|--claim-name are only valid for --type option"))
+		}
+
+		if len(a.Source) != 1 {
+			errList = append(errList, errors.New("must provide only one volume for --source"))
+		} else if len(a.Source) > 0 {
+			var source kapi.VolumeSource
+			err := json.Unmarshal([]byte(a.Source), &source)
+			if err != nil {
+				errList = append(errList, err)
+			}
+		}
+	} else if len(a.Source) > 0 || len(a.Path) > 0 || len(a.SecretName) > 0 || len(a.ClaimName) > 0 {
+		errList = append(errList, errors.New("--type|--path|--secret-name|--claim-name|--source are only valid for --add operation"))
+	}
+	return kerrors.NewAggregate(errList)
+}
+
+func (v *VolumeOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer) error {
+	clientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	v.OutputVersion = kcmdutil.OutputVersion(cmd, clientConfig.Version)
+
+	cmdNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+	mapper, typer := f.Object()
+
+	v.DefaultNamespace = cmdNamespace
+	v.Writer = out
+	v.Mapper = mapper
+	v.Typer = typer
+	v.RESTClientFactory = f.Factory.RESTClient
+	v.UpdatePodSpecForObject = f.UpdatePodSpecForObject
+	if v.Add && len(v.Name) == 0 {
+		v.Name = string(util.NewUUID())
+	}
+	return nil
+}
+
+func (v *VolumeOptions) RunVolume(args []string) error {
+	b := resource.NewBuilder(v.Mapper, v.Typer, resource.ClientMapperFunc(v.RESTClientFactory)).
+		ContinueOnError().
+		NamespaceParam(v.DefaultNamespace).DefaultNamespace().
+		SelectorParam(v.Selector).
+		ResourceTypeOrNameArgs(v.All, args...).
+		Flatten()
+
+	one := false
+	infos, err := b.Do().IntoSingular(&one).Infos()
+	if err != nil {
+		return err
+	}
+
+	// only apply resource version locking on a single resource
+	if !one && len(v.ResourceVersion) > 0 {
```

Comment:
Let's just drop that feature, it's not very useful right now and I'd rather not get sidetracked

> On Jun 9, 2015, at 8:17 PM, Andy Goldstein <notifications@github.com> wrote:
> 
> In pkg/cmd/cli/cmd/volume.go:
> 
> > +func (v *VolumeOptions) RunVolume(args []string) error {
> > +	b := resource.NewBuilder(v.Mapper, v.Typer, resource.ClientMapperFunc(v.RESTClientFactory)).
> > +		ContinueOnError().
> > +		NamespaceParam(v.DefaultNamespace).DefaultNamespace().
> > +		SelectorParam(v.Selector).
> > +		ResourceTypeOrNameArgs(v.All, args...).
> > +		Flatten()
> > +
> > +	one := false
> > +	infos, err := b.Do().IntoSingular(&one).Infos()
> > +	if err != nil {
> > +		return err
> > +	}
> > +
> > +	// only apply resource version locking on a single resource
> > +	if !one && len(v.ResourceVersion) > 0 {
> You don't appear to do anything with v.ResourceVersion other than make sure it's only used with 1 resource. Also, is the intended use of this that you'd oc get <resource>, make note of its resource version, then run oc volume --resource-version=...?
> 
> —
> Reply to this email directly or view it on GitHub.
>